plugins/inputs/vsphere/panic_handler.go

```diff
@@ -0,0 +1,28 @@
+package vsphere
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/influxdata/telegraf"
+)
+
+func HandlePanicWithAcc(acc telegraf.Accumulator) {
+	if p := recover(); p != nil {
+		switch p.(type) {
+		case string:
+			acc.AddError(errors.New(p.(string)))
+		case error:
+			acc.AddError(p.(error))
+		default:
+			acc.AddError(fmt.Errorf("Unknown panic: %s", p))
+		}
+	}
+}
+
+func HandlePanic() {
+	if p := recover(); p != nil {
+		log.Printf("E! [input.vsphere] PANIC (recovered): %s", p)
+	}
+}
```

Comment:
I added this because much of the data collection happens in separate go-routines, which, if they panic, will crash the entire process if the panic isn't recovered. I can remove this, but there's a chance panics will crash the entire Telegraf process.

Reply:
Some more detail: Telegraf gracefully handles panics in the main thread, which is great. However, panics in goroutines will crash Telegraf for the reason that's illustrated in this example: https://play.golang.org/p/jQHzZm5zMQB

This is caused by the fact that panic handlers don't cross goroutine boundaries.

I get that you want to handle panics in a single place so you can log them correctly, but I also would like to make sure Telegraf doesn't crash due to temporary glitches. The reason I implemented this in the first place was a strange, very intermittent problem related to vCenter. It's supposed to return two arrays that are guaranteed to have the same length. At some point, the arrays came back with one being longer than the other, which caused a panic. Of course, I should have checked the length and not assumed that they would be equal, but this illustrates the problem with intermittent panics that are hard to predict and that should, in my opinion, be handled in a graceful and recoverable way.

OK. I'll get off the soap box now. If you want me to toss out the panic handler, I will do so.

Reply:
Recovery in many cases isn't really possible, due to resource leaks and there will always be goroutines that you don't control in libraries.  I also don't want to hide programming errors by handling them.  I think Go best practice here is to only handle panics if you are expecting them to occur for some reason.

I'm also torn on if I should change the existing panic handling to log where to report bugs and then re-panic to crash Telegraf, or even just remove it altogether.  For now I've left it as is, but I remember noticing that it doesn't actually work very well.

So yeah, could you please remove the panic handling?

Reply:
OK. What you're saying made sense. My concern was consistency, since removing the panic handler means that some parts of the code are "safe" from panics and some aren't (i.e. code that runs in goroutines).

Anyway, I'll drop the handling now. Stand by!

Reply:
Done!